src/vm/scanning.rs

```diff
@@ -20,11 +27,11 @@ pub trait Scanning<VM: VMBinding> {
     /// in favor of bulk scanning `scan_objects`.
     ///
     /// Arguments:
-    /// * `trace`: The `TransitiveClosure` to use for scanning.
+    /// * `edge_visitor`: Called back for each edge.
     /// * `object`: The object to be scanned.
     /// * `tls`: The GC worker thread that is doing this tracing.
-    fn scan_object<T: TransitiveClosure>(
-        trace: &mut T,
+    fn scan_object<EV: EdgeVisitor>(
```

Comment:
I would suggest we change and clean up for `scan_objects()` as well. I would think the signature for `scan_objects()` should be the same as `scan_object()`, except that we use `&[ObjectReference]` instead of `ObjectReference`. 

There are multiple issues with `scan_objects()` (some are related with the changes in this PR, some are not):
* `scan_objects()` should take `EdgeVisitor` as type parameter rather than `ProcessEdgesWork`. Actually in OpenJDK's `scan_objects()` implementation, it simply creates its own `ObjectsClosure` and call `scan_object()` (which expects `EdgeVisitor`).
* OpenJDK does not need to create its own `ObjectsClosure`, and it can use the `ObjectsClosure` from mmtk-core. Actually OpenJDK first introduced `ObjectsClosure`, then `ObjectsClosure` was added to mmtk-core as V8 would use it as well. However, we never change that in OpenJDK. I am mentioning this as you also changed that part of the code in the OpenJDK binding.
* The parameter `tls` in `scan_object()` is inconsistent with `worker` in `scan_objects()`. I think we can get `tls` from `worker` but not the other way around. We probably can change `tls` in `scan_object()` to `worker` as well. I think @udesou has this change in his mmtk-core fork, and would like to upstream this change at some point as well.

As you are changing code around this, I would suggest trying to fix these issues as well. Maybe try include these changes if you find them related or easy enough to do.

Reply:
Can you remove the comments that says `scan_object()` may be removed in the future? I think I wrote that, and it is incorrect. We probably will keep both. Generally, we would use `scan_objects()`. But for some policies like Immix, we would also need to do some extra work for each scanned object (like marking lines), we would have to use `scan_object()`.

Reply:
>   I would suggest we change and clean up for scan_objects() as well. I would think the signature for scan_objects() should be the same as scan_object(), except that we use &[ObjectReference] instead of ObjectReference.

Yes. I agree. I also think all functions that can only be called by GC worker threads should take `&mut GCWorker` as the first parameter. `GCWorker` has a `tls: VMWorkerThread` field, and `VMWorkerThread` is only meaningful when calling C functions.

Reply:
> There are multiple issues with scan_objects() (some are related with the changes in this PR, some are not):
>
>    scan_objects() should take EdgeVisitor as type parameter rather than ProcessEdgesWork. Actually in OpenJDK's scan_objects() implementation, it simply creates its own ObjectsClosure and call scan_object() (which expects EdgeVisitor).
>    OpenJDK does not need to create its own ObjectsClosure, and it can use the ObjectsClosure from mmtk-core. Actually OpenJDK first introduced ObjectsClosure, then ObjectsClosure was added to mmtk-core as V8 would use it as well. However, we never change that in OpenJDK. I am mentioning this as you also changed that part of the code in the OpenJDK binding.

I am surprised that the OpenJDK binding did not fork the work packet into multiple work packets.  I would assume that if `objects.size()` is too large, it will split it into multiple work packets.  After a second thought, it can't, because the openjdk binding is not privileged to create the private `ScanObjects` work packet.

I think it is worth refactoring `scan_objects` so that it is the `mmtk-core`'s responsibility to create the `ProcessEdgesWork` work packet.

Reply:
> I am surprised that the OpenJDK binding did not fork the work packet into multiple work packets. I would assume that if `objects.size()` is too large, it will split it into multiple work packets. After a second thought, it can't, because the openjdk binding is not privileged to create the private `ScanObjects` work packet.

I guess we will not see an `ObjectReference` slice larger than `ProcessEdgesWork::CAPACITY`. So it should be fine. Plus creating more packets has the overhead of dispatching them.

> I think it is worth refactoring `scan_objects` so that it is the `mmtk-core`'s responsibility to create the `ProcessEdgesWork` work packet.

I am not exactly sure what you were referring to. But generally I find it confusing that some work is created from the binding, and it is not obvious if we only look at the mmtk-core code and try figure out where the packet is created. So it is probably a good idea if we move some packet creation back to mmtk-core. However, we would still allow a binding to create work if they want to.

Reply:
I made changes according to your suggestions.  However, I couldn't change `tls: VMWorkerThread` to `&mut GCWorker<VM>` because of some borrowing issue. It will break the following simple code:
```rust
let mut closure = ObjectsClosure::<E>::new(worker);  // NOTE: first mutable borrow here
<E::VM as VMBinding>::VMScanning::scan_objects(worker, &self.buffer, &mut closure);  // ERROR: worker was mutably borrowed.
```

Rust doesn't like a reference to be borrowed in two places. That is actually right, because if it were borrowed twice, both `ObjectClosure` and `scan_objects` could modify `worker` and it is easy to make mistakes like "changing a collection while iterating through it".

The ideal way to do this kind of callback in Rust is to add a `&mut GCWorker` parameter to `EdgeVisitor::visit_edge`, too, and let `scan_objects` pass the `&mut Worker` back.  But I think it is overly complicated.  In fact, neither OpenJDK nor JikesRVM used the `tls: VMWorkerThread` when scanning an object (JikesRVM uses it for debug purpose).  It doesn't need the Rust-level `GCWorker` structure, either.

So I made both `scan_object` and `scan_objects` take `tls: VMWorkerThread` as parameter for the following reasons:
1.  mmtk-core does not need to access any data in `GCWorekr` to support scanning object. It only uses `GCWorker` to handle the *result* of scanning objects (i.e. the edges).
2.  If any VM really needs the TLS to scan objects, we just give the TLS to the VM.